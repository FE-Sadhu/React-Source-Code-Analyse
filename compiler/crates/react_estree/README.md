# react_estree

This crate is a Rust representation of the [ESTree format](https://github.com/estree/estree/tree/master) and
popular extenions including JSX and (eventually) Flow and TypeScript.

This crate is intended as the main interchange format with outside code. A typical integration with React Compiler
will look as follows:

1. Host Compiler parses into the host AST format.
2. Host Compiler converts into `react_estree`.
3. Host Compiler invokes React Compiler to compile the input, which (conceptually) 
   returns the resulting code in `react_estree` format.
4. Host Compiler convert back from `react_estree` to its host AST format.

Because React Compiler is intended to support JavaScript-based toolchains, `react_estree` is designed to support
accurate serialization to/from estree-compatible JSON. We may also support the Babel AST format 
(a variant of ESTree) as well, depending on demand.