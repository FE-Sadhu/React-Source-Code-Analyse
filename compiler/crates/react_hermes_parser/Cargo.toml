[package]
name = "react_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
react_diagnostics = { workspace = true }
react_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
react_estree_codegen = { workspace = true }