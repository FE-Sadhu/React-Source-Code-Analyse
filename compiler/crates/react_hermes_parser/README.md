# react_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `react_estree` AST.