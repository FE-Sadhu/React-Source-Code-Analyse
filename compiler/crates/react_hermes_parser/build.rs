/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use react_estree_codegen::estree_hermes;

// Example custom build script.
fn main() {
    // Re-run if the codegen files change
    println!("cargo:rerun-if-changed=../react_estree_codegen/src/codegen.rs");
    println!("cargo:rerun-if-changed=../react_estree_codegen/src/lib.rs");
    println!("cargo:rerun-if-changed=../react_estree_codegen/src/ecmascript.json");
    println!("cargo:rerun-if-changed=../react_estree_codegen");

    let src = estree_hermes();
    let copyright = "
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
    "
    .to_string();
    let trimmed_copyright = copyright.trim();
    let contents = format!("{trimmed_copyright}\n{src}");
    std::fs::write("src/generated.rs", contents).unwrap();
}
