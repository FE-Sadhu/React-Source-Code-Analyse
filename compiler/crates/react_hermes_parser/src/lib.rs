/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

mod generated;
mod generated_extension;

use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;
use react_diagnostics::Diagnostic;
use react_estree::Program;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new(&buf);
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
