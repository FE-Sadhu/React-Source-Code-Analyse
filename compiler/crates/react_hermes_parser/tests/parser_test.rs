/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::env;

use insta::{assert_snapshot, glob};
use react_estree::SourceType;
use react_hermes_parser::parse;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let mut ast = parse(&input, path.to_str().unwrap()).unwrap();
        // TODO: hack to prevent changing lots of fixtures all at once
        ast.source_type = SourceType::Script;
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
