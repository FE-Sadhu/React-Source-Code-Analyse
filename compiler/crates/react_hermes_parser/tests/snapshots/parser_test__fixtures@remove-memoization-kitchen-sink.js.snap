---
source: crates/react_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/react_hermes_parser/tests/fixtures/remove-memoization-kitchen-sink.js
---
Input:
// @disableAllMemoization true
function Component(props) {
  const [x, setX] = useState(() => initializeState(props));
  const onChange = useCallback((e) => {
    setX(e.target.value);
  });
  const object = { x, onChange };
  return useMemo(() => {
    const { x, onChange } = object;
    return <input value={x} onChange={onChange} />;
  }, [x]);
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Component",
        "typeAnnotation": null,
        "loc": null,
        "range": [
          40,
          49
        ]
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": [
            50,
            55
          ]
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "ArrayPattern",
                  "elements": [
                    {
                      "type": "Identifier",
                      "name": "x",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": [
                        68,
                        69
                      ]
                    },
                    {
                      "type": "Identifier",
                      "name": "setX",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": [
                        71,
                        75
                      ]
                    }
                  ],
                  "loc": null,
                  "range": [
                    67,
                    76
                  ]
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "Identifier",
                    "name": "useState",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": [
                      79,
                      87
                    ]
                  },
                  "arguments": [
                    {
                      "type": "ArrowFunctionExpression",
                      "id": null,
                      "params": [],
                      "body": {
                        "type": "CallExpression",
                        "callee": {
                          "type": "Identifier",
                          "name": "initializeState",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": [
                            94,
                            109
                          ]
                        },
                        "arguments": [
                          {
                            "type": "Identifier",
                            "name": "props",
                            "typeAnnotation": null,
                            "loc": null,
                            "range": [
                              110,
                              115
                            ]
                          }
                        ],
                        "loc": null,
                        "range": [
                          94,
                          116
                        ]
                      },
                      "generator": true,
                      "async": false,
                      "loc": null,
                      "range": [
                        88,
                        116
                      ],
                      "expression": true,
                      "loc": null,
                      "range": [
                        88,
                        116
                      ]
                    }
                  ],
                  "loc": null,
                  "range": [
                    79,
                    117
                  ]
                },
                "loc": null,
                "range": [
                  67,
                  117
                ]
              }
            ],
            "loc": null,
            "range": [
              61,
              118
            ]
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "onChange",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": [
                    127,
                    135
                  ]
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "Identifier",
                    "name": "useCallback",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": [
                      138,
                      149
                    ]
                  },
                  "arguments": [
                    {
                      "type": "ArrowFunctionExpression",
                      "id": null,
                      "params": [
                        {
                          "type": "Identifier",
                          "name": "e",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": [
                            151,
                            152
                          ]
                        }
                      ],
                      "body": {
                        "type": "BlockStatement",
                        "body": [
                          {
                            "type": "ExpressionStatement",
                            "expression": {
                              "type": "CallExpression",
                              "callee": {
                                "type": "Identifier",
                                "name": "setX",
                                "typeAnnotation": null,
                                "loc": null,
                                "range": [
                                  163,
                                  167
                                ]
                              },
                              "arguments": [
                                {
                                  "type": "MemberExpression",
                                  "object": {
                                    "type": "MemberExpression",
                                    "object": {
                                      "type": "Identifier",
                                      "name": "e",
                                      "typeAnnotation": null,
                                      "loc": null,
                                      "range": [
                                        168,
                                        169
                                      ]
                                    },
                                    "property": {
                                      "type": "Identifier",
                                      "name": "target",
                                      "typeAnnotation": null,
                                      "loc": null,
                                      "range": [
                                        170,
                                        176
                                      ]
                                    },
                                    "computed": false,
                                    "loc": null,
                                    "range": [
                                      168,
                                      176
                                    ]
                                  },
                                  "property": {
                                    "type": "Identifier",
                                    "name": "value",
                                    "typeAnnotation": null,
                                    "loc": null,
                                    "range": [
                                      177,
                                      182
                                    ]
                                  },
                                  "computed": false,
                                  "loc": null,
                                  "range": [
                                    168,
                                    182
                                  ]
                                }
                              ],
                              "loc": null,
                              "range": [
                                163,
                                183
                              ]
                            },
                            "directive": null,
                            "loc": null,
                            "range": [
                              163,
                              184
                            ]
                          }
                        ],
                        "loc": null,
                        "range": [
                          157,
                          188
                        ]
                      },
                      "generator": false,
                      "async": false,
                      "loc": null,
                      "range": [
                        150,
                        188
                      ],
                      "expression": false,
                      "loc": null,
                      "range": [
                        150,
                        188
                      ]
                    }
                  ],
                  "loc": null,
                  "range": [
                    138,
                    189
                  ]
                },
                "loc": null,
                "range": [
                  127,
                  189
                ]
              }
            ],
            "loc": null,
            "range": [
              121,
              190
            ]
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "object",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": [
                    199,
                    205
                  ]
                },
                "init": {
                  "type": "ObjectExpression",
                  "properties": [
                    {
                      "type": "Property",
                      "key": {
                        "type": "Identifier",
                        "name": "x",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": [
                          210,
                          211
                        ]
                      },
                      "value": {
                        "type": "Identifier",
                        "name": "x",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": [
                          210,
                          211
                        ]
                      },
                      "kind": "init",
                      "method": false,
                      "shorthand": true,
                      "computed": false,
                      "loc": null,
                      "range": [
                        210,
                        211
                      ]
                    },
                    {
                      "type": "Property",
                      "key": {
                        "type": "Identifier",
                        "name": "onChange",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": [
                          213,
                          221
                        ]
                      },
                      "value": {
                        "type": "Identifier",
                        "name": "onChange",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": [
                          213,
                          221
                        ]
                      },
                      "kind": "init",
                      "method": false,
                      "shorthand": true,
                      "computed": false,
                      "loc": null,
                      "range": [
                        213,
                        221
                      ]
                    }
                  ],
                  "loc": null,
                  "range": [
                    208,
                    223
                  ]
                },
                "loc": null,
                "range": [
                  199,
                  223
                ]
              }
            ],
            "loc": null,
            "range": [
              193,
              224
            ]
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "CallExpression",
              "callee": {
                "type": "Identifier",
                "name": "useMemo",
                "typeAnnotation": null,
                "loc": null,
                "range": [
                  234,
                  241
                ]
              },
              "arguments": [
                {
                  "type": "ArrowFunctionExpression",
                  "id": null,
                  "params": [],
                  "body": {
                    "type": "BlockStatement",
                    "body": [
                      {
                        "type": "VariableDeclaration",
                        "kind": "const",
                        "declarations": [
                          {
                            "type": "VariableDeclarator",
                            "id": {
                              "type": "ObjectPattern",
                              "properties": [
                                {
                                  "type": "Property",
                                  "key": {
                                    "type": "Identifier",
                                    "name": "x",
                                    "typeAnnotation": null,
                                    "loc": null,
                                    "range": [
                                      262,
                                      263
                                    ]
                                  },
                                  "value": {
                                    "type": "Identifier",
                                    "name": "x",
                                    "typeAnnotation": null,
                                    "loc": null,
                                    "range": [
                                      262,
                                      263
                                    ]
                                  },
                                  "kind": "init",
                                  "computed": false,
                                  "shorthand": true,
                                  "method": false,
                                  "loc": null,
                                  "range": [
                                    262,
                                    263
                                  ]
                                },
                                {
                                  "type": "Property",
                                  "key": {
                                    "type": "Identifier",
                                    "name": "onChange",
                                    "typeAnnotation": null,
                                    "loc": null,
                                    "range": [
                                      265,
                                      273
                                    ]
                                  },
                                  "value": {
                                    "type": "Identifier",
                                    "name": "onChange",
                                    "typeAnnotation": null,
                                    "loc": null,
                                    "range": [
                                      265,
                                      273
                                    ]
                                  },
                                  "kind": "init",
                                  "computed": false,
                                  "shorthand": true,
                                  "method": false,
                                  "loc": null,
                                  "range": [
                                    265,
                                    273
                                  ]
                                }
                              ],
                              "loc": null,
                              "range": [
                                260,
                                275
                              ]
                            },
                            "init": {
                              "type": "Identifier",
                              "name": "object",
                              "typeAnnotation": null,
                              "loc": null,
                              "range": [
                                278,
                                284
                              ]
                            },
                            "loc": null,
                            "range": [
                              260,
                              284
                            ]
                          }
                        ],
                        "loc": null,
                        "range": [
                          254,
                          285
                        ]
                      },
                      {
                        "type": "ReturnStatement",
                        "argument": {
                          "type": "JSXElement",
                          "openingElement": {
                            "type": "JSXOpeningElement",
                            "name": {
                              "type": "JSXIdentifier",
                              "name": "input",
                              "loc": null,
                              "range": [
                                298,
                                303
                              ]
                            },
                            "attributes": [
                              {
                                "type": "JSXAttribute",
                                "name": {
                                  "type": "JSXIdentifier",
                                  "name": "value",
                                  "loc": null,
                                  "range": [
                                    304,
                                    309
                                  ]
                                },
                                "value": {
                                  "type": "JSXExpressionContainer",
                                  "expression": {
                                    "type": "Identifier",
                                    "name": "x",
                                    "typeAnnotation": null,
                                    "loc": null,
                                    "range": [
                                      311,
                                      312
                                    ]
                                  },
                                  "loc": null,
                                  "range": [
                                    310,
                                    313
                                  ]
                                },
                                "loc": null,
                                "range": [
                                  304,
                                  313
                                ]
                              },
                              {
                                "type": "JSXAttribute",
                                "name": {
                                  "type": "JSXIdentifier",
                                  "name": "onChange",
                                  "loc": null,
                                  "range": [
                                    314,
                                    322
                                  ]
                                },
                                "value": {
                                  "type": "JSXExpressionContainer",
                                  "expression": {
                                    "type": "Identifier",
                                    "name": "onChange",
                                    "typeAnnotation": null,
                                    "loc": null,
                                    "range": [
                                      324,
                                      332
                                    ]
                                  },
                                  "loc": null,
                                  "range": [
                                    323,
                                    333
                                  ]
                                },
                                "loc": null,
                                "range": [
                                  314,
                                  333
                                ]
                              }
                            ],
                            "selfClosing": true,
                            "loc": null,
                            "range": [
                              297,
                              336
                            ]
                          },
                          "children": [],
                          "closingElement": null,
                          "loc": null,
                          "range": [
                            297,
                            336
                          ]
                        },
                        "loc": null,
                        "range": [
                          290,
                          337
                        ]
                      }
                    ],
                    "loc": null,
                    "range": [
                      248,
                      341
                    ]
                  },
                  "generator": false,
                  "async": false,
                  "loc": null,
                  "range": [
                    242,
                    341
                  ],
                  "expression": false,
                  "loc": null,
                  "range": [
                    242,
                    341
                  ]
                },
                {
                  "type": "ArrayExpression",
                  "elements": [
                    {
                      "type": "Identifier",
                      "name": "x",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": [
                        344,
                        345
                      ]
                    }
                  ],
                  "loc": null,
                  "range": [
                    343,
                    346
                  ]
                }
              ],
              "loc": null,
              "range": [
                234,
                347
              ]
            },
            "loc": null,
            "range": [
              227,
              348
            ]
          }
        ],
        "loc": null,
        "range": [
          57,
          350
        ]
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": [
        31,
        350
      ],
      "loc": null,
      "range": [
        31,
        350
      ]
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": [
    31,
    350
  ]
}
